checksum.go

```diff
@@ -0,0 +1,245 @@
+func checksumFromFile(checksumFile string, src *url.URL) (*fileChecksum, error) {
+	checksumFileURL, err := urlhelper.Parse(checksumFile)
+	if err != nil {
+		return nil, err
+	}
+
+	tempfile, err := tmpFile("", filepath.Base(checksumFileURL.Path))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		os.Remove(tempfile)
+	}()
+
+	if err = GetFile(tempfile, checksumFile); err != nil {
+		return nil, fmt.Errorf(
+			"Error downloading checksum file: %s", err)
+	}
+
+	filename := filepath.Base(src.Path)
+	absPath, _ := filepath.Abs(src.Path)
+	relpath, _ := filepath.Rel(filepath.Dir(checksumFileURL.Path), absPath)
+
+	// possible file identifiers:
+	options := []string{
+		filename,       // ubuntu-14.04.1-server-amd64.iso
+		"*" + filename, // *ubuntu-14.04.1-server-amd64.iso  Standard checksum
+		"?" + filename, // ?ubuntu-14.04.1-server-amd64.iso  shasum -p
+		relpath,        // dir/ubuntu-14.04.1-server-amd64.iso
+		"./" + relpath, // ./dir/ubuntu-14.04.1-server-amd64.iso
+		absPath,        // fullpath; set if local
+	}
+
+	f, err := os.Open(tempfile)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"Error opening downloaded file: %s", err)
+	}
+	defer f.Close()
+	rd := bufio.NewReader(f)
+	for {
+		line, err := rd.ReadString('\n')
+		if err != nil {
+			if err != io.EOF {
+				return nil, fmt.Errorf(
+					"Error reading checksum file: %s", err)
+			}
+			break
+		}
+		checksum, err := parseChecksumLine(line)
+		if err != nil || checksum == nil {
+			continue
+		}
+		if checksum.Filename == "" {
+			// filename not sure, let's try
+			return checksum, nil
+		}
+		// make sure the checksum is for the right file
+		for _, option := range options {
+			if checksum.Filename == option {
+				// any checksum will work so we return the first one
+				return checksum, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("no checksum found in: %s", checksumFile)
+}
+
+// parseChecksumLine takes a line from a checksum file and returns
+// checksumType, checksumValue and filename parseChecksumLine guesses the style
+// of the checksum BSD vs GNU by splitting the line and by counting the parts.
+// of a line.
+// for BSD type sums parseChecksumLine guesses the hashing algorithm
+// by checking the length of the checksum.
+func parseChecksumLine(line string) (*fileChecksum, error) {
+	parts := strings.Fields(line)
+
+	switch len(parts) {
+	case 4:
+		// BSD-style checksum:
+		//  MD5 (file1) = <checksum>
+		//  MD5 (file2) = <checksum>
+		if len(parts[1]) <= 2 ||
+			parts[1][0] != '(' || parts[1][len(parts[1])-1] != ')' {
+			return nil, fmt.Errorf(
+				"Unexpected BSD-style-checksum filename format: %s", line)
+		}
+		filename := parts[1][1 : len(parts[1])-1]
+		return newChecksumFromType(parts[0], parts[3], filename)
+	case 2:
+		// GNU-style:
+		//  <checksum>  file1
+		//  <checksum> *file2
+		return newChecksumFromValue(parts[0], parts[1])
+	case 0:
+		return nil, nil // empty line
+	default:
+		return newChecksumFromValue(parts[0], "")
```

Comment:
I can't tell if this is covered by tests?

Reply:
I did add some tests in this file. ex: 
https://github.com/hashicorp/go-getter/blob/487c31a11da44526b399b80b2401c63ce5bc5e1d/get_test.go#L291-L295

Should I also test for weird cases ?

Reply:
Oh right ! I added a test in 724c0bcbca75a183483d63e5a904891d8e8a12fc